### ISSUE TITLE
Inheritance 4

### ISSUE THREAD

**Issue by Udayv56** [Udayv56/CPP#26]
Multilevel inheritance is a feature of object-oriented programming where a derived class is derived from another derived class. This creates a chain or hierarchy of classes. Each derived class inherits the members (attributes and methods) of its immediate base class, and it can also add its own members or override inherited members.

**CODE-**
#include<iostream>
using namespace std;
// Multilevel inheritance
class Student{
    protected:
        int roll_number;
    public:
        void setNum(int);
        void getNum(void);
};

void Student :: setNum(int r){
    roll_number = r;
}

void Student :: getNum(){
    cout<<roll_number<<endl;
}

class Exam : public Student{
    protected:
        float maths;
        float physics;
    public:
        void setMarks(float, float);
        void getMarks(void);
};

void Exam :: setMarks(float m1 ,float m2){
    maths = m1;
    physics = m2;
}

void Exam :: getMarks(){
    
    cout<<maths<<" "<<physics<<endl;
}

class Result : public Exam{
    float pert;
    public:
        void display();
};

void Result :: display(){
    getNum();
    getMarks();
    
    cout<<(maths+physics)/2<<endl;
}

int main(){
    Result r;
    r.setNum(2019245);
    r.setMarks(96.0,86.0);
    r.getMarks();
    r.display();
    return 0;
}

**OUTPUT-**
96 86
2019245
96 86
91

Let's go through the provided code step by step:

1. The code begins with the inclusion of the necessary header file, `<iostream>`, for input/output operations.

2. The `Student` class is defined, which has one protected data member `roll_number`. It also has two member functions: `setNum()` to set the roll number and `getNum()` to get the roll number.

3. The `setNum()` function is defined outside the class. It sets the value of `roll_number` to the provided value.

4. The `getNum()` function is defined outside the class. It prints the value of `roll_number` to the console.

5. The `Exam` class is defined, which is derived from the `Student` class using the public access specifier. It adds two protected data members: `maths` and `physics`. It also has two member functions: `setMarks()` to set the marks in maths and physics, and `getMarks()` to get the marks in maths and physics.

6. The `setMarks()` function is defined outside the class. It sets the values of `maths` and `physics` to the provided values.

7. The `getMarks()` function is defined outside the class. It prints the values of `maths` and `physics` to the console.

8. The `Result` class is defined, which is derived from the `Exam` class using the public access specifier. It adds one private data member `pert` (percentage). It also has one member function `display()`.

9. The `display()` function is defined outside the class. It first calls the `getNum()` function inherited from the `Student` class to print the roll number. Then, it calls the `getMarks()` function inherited from the `Exam` class to print the marks in maths and physics. Finally, it calculates the average marks by adding `maths` and `physics` and dividing the sum by 2. It prints the calculated average to the console.

10. In the `main()` function, an object `r` of the `Result` class is created.

11. The `setNum()` function is called on the object `r` to set the roll number.

12. The `setMarks()` function is called on the object `r` to set the marks in maths and physics.

13. The `getMarks()` function is called on the object `r` to print the marks in maths and physics.

14. The `display()` function is called on the object `r` to display the roll number, marks, and average marks.

15. The program ends with a `return 0` statement.

In summary, this code demonstrates multilevel inheritance, where the `Result` class is derived from the `Exam` class, which in turn is derived from the `Student` class. It shows how the derived class inherits the members from its base classes and how it can access and utilize those members.
